juju/juju
# Validate that resources returned from Charmhub API match metadata.yaml

Sometimes the Charmhub API doesn't return resources (either due to a data modelling issue on the Charmhub API server side, or because of an old version of charmcraft that didn't check this?), see https://bugs.launchpad.net/juju/+bug/1980249 (and API response example [here](https://pastebin.canonical.com/p/Nj59hWZr2y/)).

Add validation when we first get the "essential metadata" before deploy, that ensures all the resources in `metadata.yaml` are present in the Charmhub API response for that revision of the charm. Note that we only do this when requesting by channel, not revision, because resources are specific to a channel and Charmhub doesn't return resources when reqeusting by revision (unless you ask for specific resource revisions).

## Checklist


## QA steps

The API response for that charm (`prometheus2`) is now correct -- that is, now returning the zero-length resources correctly. So to reproduce the issue you have to hack `charmhub/refresh.go` to set the `Resources` field to an empty slice as follows:

```
diff --git a/charmhub/refresh.go b/charmhub/refresh.go
index 7a179c4c60..d227e878fe 100644
--- a/charmhub/refresh.go
+++ b/charmhub/refresh.go
@@ -177,6 +177,10 @@ func (c *refreshClient) refresh(ctx context.Context, ensure func(responses []tra
        if err := ensure(resp.Results); err != nil {
                return nil, errors.Trace(err)
        }
+       for i := range resp.Results {
+               resp.Results[i].Entity.Resources = nil // TODO
+       }
+       c.logger.Errorf("TODO refreshClient.refresh:\n%s", pretty.Sprint(resp.Results))
        // Exit early.
        if len(resp.Results) <= 1 {
                return resp.Results, nil
```

Then you just bootstrap (I'm using LXD), and deploy. It will deploy successfully, but then in the logs you'll see the warning similar to the one in the bug (`while getting resource from charmhub: resource "core" not found`).

Before the PR:

```sh
# apply patch above
$ juju bootstrap lxd
$ juju add-model t
$ juju deploy prometheus2
$ juju debug-log -m controller
# look for warning in log
```

After this PR fix,

```sh
# same steps as above, but the deploy should error:
$ juju deploy prometheus2
ERROR storing charm "prometheus2": retrieving essential metadata for charm "ch:amd64/focal/prometheus2-30": \
  validating resources in Charmhub API response for "ch:amd64/focal/prometheus2-30", revision 30: \
  missing core, prometheus
```

## Bug reference

https://bugs.launchpad.net/juju/+bug/1980249
